nixos/modules/services/web-apps/kasmweb.nix

```diff
@@ -0,0 +1,342 @@
+
+let
+  cfg = config.services.kasmweb;
+in
+{
+  options.services.kasmweb = {
+    enable = mkEnableOption (lib.mdDoc "kasmweb");
+
+    networkSubnet = mkOption {
+      default = "172.20.0.0/16";
+      type = types.str;
+      description = lib.mdDoc ''
+        The network subnet to use for the containers.
+      '';
+    };
+
+    postgres = {
+      user = mkOption {
+        default = "kasmweb";
+        type = types.str;
+        description = lib.mdDoc ''
+          Username to use for the postgres database.
+        '';
+      };
+      password = mkOption {
+        default = "kasmweb";
+        type = types.str;
+        description = lib.mdDoc ''
+          password to use for the postgres database.
+        '';
+      };
+    };
+
+    redisPassword = mkOption {
+      default = "kasmweb";
+      type = types.str;
+      description = lib.mdDoc ''
+        password to use for the redis cache.
+      '';
+    };
+
+    defaultAdminPassword = mkOption {
+      default = "kasmweb";
+      type = types.str;
+      description = lib.mdDoc ''
+        default admin password to use.
+      '';
+    };
+
+    defaultUserPassword = mkOption {
+      default = "kasmweb";
+      type = types.str;
+      description = lib.mdDoc ''
+        default user password to use.
+      '';
+    };
+
+    defaulManagerToken = mkOption {
+      default = "kasmweb";
+      type = types.str;
+      description = lib.mdDoc ''
+        default manager token to use.
+      '';
+    };
+
+    defaulGuacToken = mkOption {
+      default = "kasmweb";
+      type = types.str;
+      description = lib.mdDoc ''
+        default guac token to use.
+      '';
+    };
+
+    defaulRegistrationToken = mkOption {
+      default = "kasmweb";
+      type = types.str;
+      description = lib.mdDoc ''
+        default registration token to use.
+      '';
+    };
+
+    datastorePath = mkOption {
+      type = types.str;
+      default = "/var/lib/kasmweb";
+      description = lib.mdDoc ''
+        The directory used to store all data for kasmweb.
+      '';
+    };
+
+    sslCertificate = mkOption {
+      type = types.nullOr types.path;
+      default = null;
+      description = lib.mdDoc ''
+        The SSL certificate to be used for kasmweb.
+      '';
+    };
+
+    sslKey = mkOption {
+      type = types.nullOr types.path;
+      default = null;
+      description = lib.mdDoc ''
+        The SSL certificate's key to be used for kasmweb. Make sure to specify
+        this as a string and not a literal path, so that it is not accidentally
+        included in your nixstore.
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+    system.activationScripts.kasmweb-make-ready = let
+      docker = config.virtualisation.oci-containers.backend;
+      dockerBin = "${pkgs.${docker}}/bin/${docker}";
+      api_hostname = "kasm_api";
+      manager_hostname = "kasm_proxy";
+      sed = "${pkgs.gnused}/bin/sed";
+    in ''
+      mkdir -p ${cfg.datastorePath}/log
```

Comment:
This is a huge script - 233-120=113 lines of Bourne Shell embedded in a Nix file, with a truckload of interpolations.

Definitely this is very hacky, it should be a better way to write all of it.

Reply:
Given that the vendor's installation script is a 1400 line bash script, I'd say this is an improvement ;)

In all seriousness though, do you have any concrete suggestions?

Reply:
This script should be somehow outsourced, living in a (say) `./make-ready.sh` file.

This way, you will need to create a subdir `services/web-apps/kasmweb/` and move `kasmweb.nix` inside it, optionally under the name `default.nix`.

The other thing to do is use `lib.substituteAll` in order to substitute the many string interpolations you did.

The next other thing is to make this `make-ready.sh` script a nix-shell script, declaring all these dependencies - openssl, docker, gnused...

Reply:
Ok. I'll start by shipping the shell script as a separate file.

Reply:
I ended up making it into a systemd oneshot service. Let me know what you think...